Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/DafnySequence.java

```diff
@@ -265,11 +332,24 @@ public boolean contains(T t) {
     }
 
     @Override
-    public boolean equals(Object obj) {
-        if (this == obj) return true;
-        if (!(obj instanceof DafnySequence<?>)) return false;
-        DafnySequence<?> o = (DafnySequence<?>) obj;
-        return asList().equals(o.asList());
+    public final boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (!(obj instanceof DafnySequence<?>)) {
+            return false;
+        }
+        @SuppressWarnings("unchecked")
+        DafnySequence<T> other = (DafnySequence<T>) obj;
+        return this.equalsNonLazy(other.force());
+    }
+
+    /**
+     * Compare for equality to the given sequence, assuming that it is not
+     * null, not == to this, and not lazy.
+     */
+    protected boolean equalsNonLazy(NonLazyDafnySequence<T> other) {
+        return asList().equals(other.asList());
```

Comment:
Same comment here as in `toArray()`

Reply:
This one is actually necessary as a fallback.  All the overrides expect an argument of a certain type, and if the argument isn't that type, we can't do any better than iterating over both sequences, which is what this line really does.